### PR TITLE
Stat override message

### PR DESCRIPTION
**Pull request by ncontinanza**
Part of https://github.com/lambdaclass/champions_of_mirra/issues/280

## Motivation

We should have a specific message for the case in which we want to override a stat's value, despite its current value.
In the immediate future this will be used to set the energy to 0 after casting the ultimate skill, in the near future it will be used to set other stats such as the health.

## Summary of changes

- Create a new StatOverride message.
- Send this message when casting a unit's ultimate skill, setting its energy to 0.

## How to test it?

You can use the SocketTester (remember to move it to the gateway directory) to simulate a battle and check that this new message is sent. You can follow these steps to do so:

```
IEx.configure(inspect: [limit: :infinity])
```

```
{:ok, user} = Champions.Users.register("Lotu")
```

```
campaign_progress = user.super_campaign_progresses |> hd()
```

```
{:ok, pid} = Gateway.SocketTester.start_link
```

```
Gateway.SocketTester.fight_level(pid, user.id, campaign_progress.level_id)
```

If you feel more comfortable using the client, start the client using [this branch](https://github.com/lambdaclass/champions_of_mirra/pull/420), go to BattleManager.cs and add a Debug.Log() in the PlayOutSteps() method, in the StatOverride case of the switch statement. Run a battle and there you should see how the frontend receives this new message when a unit is about to use the ulti.

## Checklist
- [X] Tested the changes locally.
- [X] Reviewed the changes on GitHub, line by line.
- [ ] This change requires new documentation.
  - [ ] Documentation has been added/updated.

### CODE REVIEW

**Comment by lotuuu** [comment]
Let's implement statoverrides for heal caps

File: apps/champions/lib/champions/battle/simulator.ex [hunk 1]
```diff
@@ -894,13 +894,35 @@ defmodule Champions.Battle.Simulator do
...
+            },
+            :stat_override
+          )
+        else
+          history_acc
+        end
+      end)
+
+    new_state =
+      Map.put(
```
Comment (lotuuu) [inline apps/champions/lib/champions/battle/simulator.ex:917]: We're reducing the same list twice! Let's use a tuple as the accumulator, like `{new_history, new_state} = Enum.reduce(state.units, {history, state}, fn {_unit_id, unit} -> ...`

**Review by lotuuu — CHANGES_REQUESTED** [review]

File: apps/champions/lib/champions/battle/simulator.ex [hunk 2]
```diff
@@ -195,6 +195,16 @@ defmodule Champions.Battle.Simulator do
               },
               :skill_action
             )
+            |> add_to_history(
+              %{
+                target_id: unit.id,
+                stat_affected: %{
+                  stat: :ENERGY,
+                  amount: -@ultimate_energy_cost
```
Comment (lotuuu) [inline apps/champions/lib/champions/battle/simulator.ex:203]: This should be 0, otehrwise the whole concept of stat_override doesn't make sense

**Review by lotuuu — CHANGES_REQUESTED** [review]

**Review by lotuuu — APPROVED** [review]
To test with SocketTester:
```
IEx.configure(inspect: [limit: :infinity])
``
`

```
{:ok, user} = Champions.Users.register("Lotu")
```

```
campaign_progress = user.super_campaign_progresses |> hd()
```

```
{:ok, pid} = Gateway.SocketTester.start_link
```

```
Gateway.SocketTester.fight_level(pid, user.id, campaign_progress.level_id)
```

**Comment by ncontinanza** [comment]
> To test with SocketTester:
> 
> ```
> IEx.configure(inspect: [limit: :infinity])
> ```
> 
> ```
> {:ok, user} = Champions.Users.register("Lotu")
> ```
> 
> ```
> campaign_progress = user.super_campaign_progresses |> hd()
> ```
> 
> ```
> {:ok, pid} = Gateway.SocketTester.start_link
> ```
> 
> ```
> Gateway.SocketTester.fight_level(pid, user.id, campaign_progress.level_id)
> ```

Nice! I'll add it to the description

**Review by tkz00 — APPROVED** [review]
Passed functional review!

**Pull request merged**